tests/test_oic_provider_logout.py

```diff
@@ -0,0 +1,1012 @@
+        # I know the session backend DB is a DictSessionBackend so I can use that
+        _sid = list(self.provider.sdb._db.storage.keys())[0]
+        _sub = self.provider.sdb[_sid]["sub"]
+        #
+        if "backchannel_logout_uri" in self.provider.cdb["number5"]:
+            del self.provider.cdb["number5"]["backchannel_logout_uri"]
+
+        res = self.provider.do_back_channel_logout(
+            self.provider.cdb["number5"], _sub, _sid
+        )
+        assert res is None
+
+    def test_id_token_hint_multiple_aud(self):
+        id_token = self._auth_with_id_token()
+        assert session_get(
+            self.provider.sdb, "sub", id_token["sub"]
+        )  # verify we got valid session
+
+        self.provider.cdb["number5"]["post_logout_redirect_uris"] = [
+            ("https://example.com/plru", "")
+        ]
+
+        # add another aud and an azp.
+        id_token["azp"] = id_token["aud"][0]
+        id_token["aud"].append("foobar")
+        id_token_hint = id_token.to_jwt(algorithm="none")
+
+        resp = self.provider.end_session_endpoint(
+            urlencode({"id_token_hint": id_token_hint})
+        )
+
+        assert isinstance(resp, SeeOther)
+
+    def test_id_token_hint_aud_does_not_match_client_id(self):
+        id_token = self._auth_with_id_token()
+        assert session_get(
+            self.provider.sdb, "sub", id_token["sub"]
+        )  # verify we got valid session
+
+        # add another aud and an azp.
+        id_token_hint = id_token.to_jwt(algorithm="none")
+
+        # Mess with the session DB
+        _sid = list(self.provider.sdb._db.storage.keys())[0]
+        self.provider.sdb[_sid]["client_id"] = "something else"
+        resp = self.provider.end_session_endpoint(
+            urlencode({"id_token_hint": id_token_hint})
+        )
+
+        assert isinstance(resp, Response)
+        assert resp.status_code == 400
+
+    def test_no_back_or_front_channel_logout(self):
+        self._code_auth()
+
+        # Mess with client DB
+        for c in ["backchannel_logout_uri", "frontchannel_logout_uri"]:
+            if c in self.provider.cdb["number5"]:
+                del self.provider.cdb["number5"][c]
+
+        resp = self.provider.do_verified_logout(
+            sid=list(self.provider.sdb._db.storage.keys())[0], client_id="number5"
+        )
+
+        assert resp == {}
+
+    def test_back_channel_http_exception(self):
+        self._code_auth()
+
+        # client0
+        self.provider.cdb["client0"][
+            "backchannel_logout_uri"
+        ] = "https://example.com/bc_logout"
+        self.provider.cdb["client0"]["client_id"] = "client0"
+
+        # Get a session ID, anyone will do.
+        # I know the session backend DB is a DictSessionBackend so I can use that
+        _sid = list(self.provider.sdb._db.storage.keys())[0]
+
+        # There is no back channel logout, hence there should be no HTTP POST
+        exception = ConnectionError()
```

Comment:
The exception is now unused, but quality will probably complain about that as well :)

Reply:
Removed